File: ReactNativeApp/ios/Modules/Phases/Instances/CalibrationPhase.swift
//
//  CalibrationPhase.swift
//  jogo
//
//  Created by Mohsin on 22/10/2021.
//

import Foundation

public class CalibrationPhase: Phase {
  
  var sensorCalibration: SSVC<SensorCalibrationState, SensorCalibrationView, SensorCalibrationController>!
  var exerciseCalibration: SSVC<ExerciseCalibrationState, ExerciseCalibrationView, ExerciseCalibrationController>!
  var calibrationVisuals: SSVC<CalibrationVisualsState, CalibrationVisualsView, CalibrationVisualsController>!
  var person: PersonSSVC<PersonState, PersonView, PersonController>!
  var ball: BallSSVC<BallState, BallView, BallController>!
  
  public override func process() {
    exerciseCalibration.controller.process()
    person.controller.changeCalibrationStatus(isCalibrated: exerciseCalibration.controller.isPersonCalibrated())
    ball.controller.changeCalibrationStatus(isCalibrated: exerciseCalibration.controller.isBallCalibrated())
    super.process()
    Profiler.get().profileTime(frameId: GameContext.getContext()?.currentFrameNumber ?? -1, tag: "\(self.phaseName ?? "" )-\(#function)", delta: Date().getMilliseconds() - timings["\(#function)"]!)
  }
  
  public override func isDone() -> Bool {
    return sensorCalibration.controller.isRotationCalibrated() && exerciseCalibration.controller.isCalibrated()
  }
  
  public override func initialize() {
    self.phaseName = "Calibration"
    self.prevPhase = .PromoPhase
    self.nextPhase = .CountDownPhase
    // Build Sensor Calibration SSVC
    sensorCalibration = SensorCalibrationBuilder.build()
    
    // Build Exercise Calibration SSVC
    exerciseCalibration = ExerciseCalibrationBuilder.build()
    
    calibrationVisuals = CalibrationVisualsBuilder.build()
    calibrationVisuals.controller.setDelegateToButton(phase: self, action: #selector(questionMarkBtnAction(_:)))
    person = PersonBuilder.build(state: GameContext.getContext()!.personState!)
    ball = BallBuilder.build(state: GameContext.getContext()!.ballState!)
    ball.changeView(drawType: .POINT)
    super.initialize()
    Profiler.get().profileTime(frameId: GameContext.getContext()?.currentFrameNumber ?? -1, tag: "\(self.phaseName ?? "" )-\(#function)", delta: Date().getMilliseconds() - timings["\(#function)"]!)
  }
  
  @objc func questionMarkBtnAction(_ sender: UIButton) {
    self.delegate?.changePhase(to: .VideoPhase)
  }
  
  public override func cleanup() {
    super.cleanup()
    self.sensorCalibration.cleanup()
    self.exerciseCalibration.cleanup()
    self.calibrationVisuals.cleanup()
    self.person.cleanup(with: false)
    self.ball.cleanup(with: false)
    self.sensorCalibration = nil
    self.exerciseCalibration = nil
    self.calibrationVisuals = nil
    self.person = nil
    self.ball = nil
    Profiler.get().profileTime(frameId: GameContext.getContext()?.currentFrameNumber ?? -1, tag: "\(self.phaseName ?? "" )-\(#function)", delta: Date().getMilliseconds() - timings["\(#function)"]!)
  }
  
  public override func getDebugText() -> String {
//    return "isSensorCalibrated: \(sensorCalibration.controller.isRotationCalibrated()) \nisExerciseCalibrated: \(exerciseCalibration.controller.isCalibrated())"
//    GameContext.getContext()!.objectDetections.forEach { objState in
//
//      (objState as? PersonState)?.getDetectionSubClasses().forEach({ subStates in
//        print("\(subStates.getLabel()), \(subStates.getX()), \(subStates.getY())")
//      })
//
//      print((objState as? BallState)?.description)
//    }
    return GameContext.getContext()!.objectDetections.description
  }
}

File: ReactNativeApp/android/app/src/main/java/com/phases/instances/CalibrationPhase.java
package com.phases.instances;

import com.activities.GameActivity;
import com.components.basic.calibration.CalibrationBuilder;
import com.components.basic.calibration.CalibrationController;
import com.components.basic.calibration.CalibrationState;
import com.components.basic.calibration.CalibrationView;
import com.components.basic.calibration.SensorCalibrationBuilder;
import com.components.basic.calibration.SensorCalibrationController;
import com.components.basic.calibration.SensorCalibrationState;
import com.components.basic.calibration.SensorCalibrationView;
import com.phases.base.Phase;
import com.phases.phasemanager.PhaseManager;
import com.ssvc.SSVC;

public class CalibrationPhase extends Phase {

    SSVC<CalibrationState, CalibrationView, CalibrationController> calibration;
    SSVC<SensorCalibrationState, SensorCalibrationView, SensorCalibrationController> sensorCalibration;

    @Override
    public void process() {
        calibration.controller.processCalibration();
        if (calibration.controller.isCalibrated())
            sensorCalibration.controller.cleanUp(false);
    }

    @Override
    public void initialize() {
        this.prevPhase = PhaseManager.GamePhases.PROMO_PHASE;
        this.nextPhase = PhaseManager.GamePhases.COUNTDOWN_PHASE;
        calibration = CalibrationBuilder.buildCalibration();
//        if (GameActivity.getActivity().isOrientationLandscape())
        sensorCalibration = SensorCalibrationBuilder.buildSensorCalibration();
        GameActivity.getActivity().showButtons();

    }

    @Override
    public boolean isDone() {
        return (sensorCalibration.controller.isRotationCalibrated() && calibration.controller.isCalibrated());
//        return calibration.controller.isCalibrated();
    }

    @Override
    public String getDebugText() {
        return "";
    }

    @Override
    public void cleanup() {
        super.cleanup();
        sensorCalibration.controller.cleanUp(true);
        sensorCalibration.state.cleanup();
        sensorCalibration.stateView.detachFromLayoutSV();
        calibration.stateView.detachFromLayoutSV();
    }

    @Override
    public PhaseManager.GamePhases getNextPhase() {
        return this.nextPhase;
    }

    @Override
    public PhaseManager.GamePhases getPrevPhase() {
        return this.prevPhase;
    }

}
